File: MacApp/MacWorkspaceFrame.vb
Imports AppKit
Imports RemObjects.Elements.RTL
Imports Core

  ' NSImage keeps the captured CGImage alive after ScreenCaptureKit returns.
  ' Shared owns this wrapper until presentFrame succeeds; the surface releases it.
Public Class MacWorkspaceFrame
  Implements IWorkspaceFrame

  Private _image As NSImage

  Public Sub New(image As NSImage)
    If image Is Null Then
      Throw New ArgumentNullException("image")
    End If

    _image = image
  End Sub

  Public ReadOnly Property IsReleased() As Boolean
    Get
      Return _image Is Null
    End Get
  End Property

  Public Function imageForPresentation() As NSImage
    Return _image
  End Function

  Public Sub releaseFrame()
    _image = Null
  End Sub

End Class

File: MacApp/MacWorkspaceFrame.pas
namespace MacApp;

uses
  AppKit,
  RemObjects.Elements.RTL,
  Core;

type
  // NSImage keeps the captured CGImage alive after ScreenCaptureKit returns.
  // Shared owns this wrapper until presentFrame succeeds; the surface releases it.
  MacWorkspaceFrame = public class(IWorkspaceFrame)
  private
    fImage: NSImage;
    method getIsReleased: Boolean;
    begin
      result := fImage = nil;
    end;
  public
    constructor(image: NSImage);
    begin
      if image = nil then
        raise new ArgumentNullException('image');

      fImage := image;
    end;
    property IsReleased: Boolean read getIsReleased;
    method imageForPresentation: NSImage;
    begin
      result := fImage;
    end;
    method releaseFrame;
    begin
      fImage := nil;
    end;
  end;

end.
